cache/memoize.ts

```diff
@@ -0,0 +1,203 @@
+ *
+ * // fibonacci function, which is very slow for n > ~30 if not memoized
+ * const fib = memoize((n: bigint): bigint => {
+ *   return n <= 2n ? 1n : fib(n - 1n) + fib(n - 2n);
+ * });
+ *
+ * assertEquals(fib(100n), 354224848179261915075n);
+ * ```
+ */
+const memoize_: typeof memoize = memoize(memoize, {
+  cache: _cache,
+  getKey(fn, options) {
+    const optionVals = Object.entries(options ?? {})
+      .sort(([a], [b]) => a > b ? 1 : a < b ? -1 : 0)
+      .map(([, v]) => v);
+    return _getKey(fn, ...optionVals);
+  },
+});
+
+export { memoize_ as memoize };
+
+function memoize<Fn extends (...args: never[]) => unknown>(
+  fn: Fn,
+  options?: NoInfer<Partial<MemoizationOptions<Fn>>>,
+): Fn & {
+  cache: MemoizationCache<unknown, ReturnType<Fn>>;
+  getKey: (this: ThisParameterType<Fn>, ...args: Parameters<Fn>) => unknown;
+} {
+  const cache = options?.cache ?? new Map();
+  const getKey = options?.getKey ?? _serializeArgList(cache);
+  const truncateArgs = options?.truncateArgs ?? false;
+  const cacheRejectedPromises = options?.cacheRejectedPromises ?? false;
+
+  const memoized = function (
+    this: ThisParameterType<Fn>,
+    ...args: Parameters<Fn>
+  ): ReturnType<Fn> {
+    if (truncateArgs) args = args.slice(0, fn.length) as Parameters<Fn>;
+
+    const key = getKey.apply(this, args);
+
+    if (cache.has(key)) {
+      return cache.get(key)!;
+    }
+
+    let val = fn.apply(this, args) as ReturnType<Fn>;
+
+    if (val instanceof Promise && !cacheRejectedPromises) {
+      val = val.catch((reason) => {
+        cache.delete(key);
+        throw reason;
+      }) as typeof val;
+    }
+
+    cache.set(key, val);
+
+    return val;
+  } as Fn;
+
+  return Object.defineProperties(Object.assign(memoized, { cache, getKey }), {
+    length: { value: fn.length },
+    name: { value: fn.name },
+  });
+}
+
+/** Default serialization of arguments list for use as cache keys */
+export function _serializeArgList<Return>(
+  cache: MemoizationCache<unknown, Return>,
+): (this: unknown, ...args: unknown[]) => string {
+  const weakKeyToKeySegmentCache = new WeakMap<WeakKey, string>();
+  const weakKeySegmentToKeyCache = new Map<string, string[]>();
+  let i = 0;
+
+  const registry = new FinalizationRegistry<string>((keySegment) => {
+    for (const key of weakKeySegmentToKeyCache.get(keySegment) ?? []) {
+      cache.delete(key);
+    }
+    weakKeySegmentToKeyCache.delete(keySegment);
+  });
+
+  return function (...args) {
+    const weakKeySegments: string[] = [];
+    const keySegments = [this, ...args].map((arg) => {
+      if (typeof arg === "undefined") return "undefined";
+      if (typeof arg === "bigint") return `${arg}n`;
+
+      if (
+        arg === null ||
+        ["string", "boolean", "number", "record", "tuple"].includes(typeof arg)
+      ) {
+        // null, string, boolean, number, or one of the upcoming
+        // [record/tuple](https://github.com/tc39/proposal-record-tuple)
+        // ECMAScript types
```

Comment:
I think it's too early to include "record" and "tuple" here.

Reply:
An alternative that would avoid explicitly handling record/tuple yet would remain forward-compatible for them would be to negate the condition. Something like:

```diff
- if (
-    arg === null ||
-    ["string", "boolean", "number", "record", "tuple"].includes(typeof arg)
- ) 
+ if (
+    typeof arg !== 'function' && (arg === null || typeof arg !== 'object')
+ )
```

Intention is less clear that way though.

Reply:
Is that safe to do? I slightly prefer only including `null`, `string`, `boolean`, and `number`, which feels easier to read.

One reason why I'd like to drop record and tuple is that I heard the rumor that proposal is not going forward anymore

Reply:
> Is that safe to do? I slightly prefer only including null, string, boolean, and number, which feels easier to read.

Makes sense, I'll change to that along with a comment about forward-compatibility.

> I heard the rumor that proposal is not going forward

Sad if true 😔